### PR TITLE
RFC 185: Add WebDriver BiDi support to testdriver.js

### PR DESCRIPTION
**Pull request by sadym-chromium**
Re-apply https://github.com/web-platform-tests/rfcs/pull/182

Add “testdriver.js” support for [WebDriver BiDi](https://w3c.github.io/webdriver-bidi/) events and actions.

***
<a href="https://github.com/sadym-chromium/wpt_rfcs/blob/sadym/testdriver-bidi/rfcs/testdriver_bidi.md" title="Latest version">Preview</a>

### CODE REVIEW

File: rfcs/testdriver_bidi.md [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+
+## Background
+
+The “testdriver.js” framework (or testdriver for simplicity) is based on the one-directional WebDriver Classic protocol and lacks the ability to receive events from the browser via wptrunner, limiting our ability to test event-based functionalities such as console events in WPT (e.g. many [console log tests](https://github.com/web-platform-tests/wpt/tree/master/console) are [manual](https://github.com/web-platform-tests/wpt/blob/master/console/console-timing-logging-manual.html)!). To overcome this limitation, this RFC proposes using the [WebDriver BiDi](https://w3c.github.io/webdriver-bidi) protocol, an extension for WebDriver Classic offering bidirectional communication between the test and the browser.
+
+This document presents changes in testdriver and wptrunner so that it incorporates both WebDriver Classic and WebDriver BiDi functionalities, demonstrating the addition of tests for console events as an [example](https://github.com/web-platform-tests/wpt/pull/44649).
```
Comment (foolip) [inline rfcs/testdriver_bidi.md:15]:
```suggestion
This RFC presents changes in testdriver and wptrunner so that it incorporates both WebDriver Classic and WebDriver BiDi functionalities, demonstrating the addition of tests for console events as an [example](https://github.com/web-platform-tests/wpt/pull/44649).
```

**Review by foolip — APPROVED** [review]
I've reviewed this in the form of a document before. Spotted one minor thing when skimming again.

**Comment by jgraham** [comment]
As discussed in the infra meeting I'm going to review this, but it may take some time since it's a rather substantial proposal that is likely to have significant knock-on effects. It would also be useful to get review from anyone who is not planning to use the webdriver backend to wptrunner in the near future; that might include WebKit (e.g. @gsnedders) and Servo (@mrobinson perhaps).

**Comment by OrKoN** [comment]
Friendly ping to review the RFC @jgraham @gsnedders. There is a meeting tonight that offers an opportunity to discuss the proposed changes if there are any questions.

**Comment by jgraham** [comment]
Just FYI, I have started looking at this. I'm not sure if it's making the right set of tradeoffs in the current form, but I need more time to look again.

**Review by jgraham — CHANGES_REQUESTED** [review]
Apologies for the long delay in providing concrete feedback here.

I'm generally very worried about the potential to leak state between tests. I think the harness needs to take responsibility for managing at least event subscription state rather than depending on developers to correctly clean up after every test.

I'm also worried about the fact that we seem to be keeping some of the hacks that were only really required because of limitations of WebDriver. In particular keeping the ActionsContext for BiDi actions seems likely to cause problems.

I'd also like to see examples of how we could use the API for use cases that currently aren't possible or are prohibitively difficult, for example running actions in cross-origin windows. BiDi should unlock that, since it adds the ability to message arbitrary contexts. However the focus on just getting console.log events doesn't really reveal if this design is a useful stepping stone to that outcome. I worry that choices here could impede our ability to make the best use of BiDi in the future, if tests end up relying on accidental features of the design.

File: rfcs/testdriver_bidi.md [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+### Example
+
+Here is an [example](https://github.com/sadym-chromium/wpt/blob/sadym/testdriver-bidi/console/console-log-logged.html) of a test using the proposed `test_driver.bidi` API for testing console log events.
+```javascript
+promise_test(async (test) => {
+    const some_message = "SOME MESSAGE";
+    // Subscribe to `log.entryAdded` BiDi events. This will not add a listener to the page.
+    await test_driver.bidi.log.entry_added.subscribe();
+    test.add_cleanup(async function() {
```
[the remark this answers is not in the document]
Comment (sadym-chromium) [reply rfcs/testdriver_bidi.md:33]: In order to address the concern, we need to add a dependency between testharness and testdriver. I see the following ways:
1. Make testharness aware of the events concept and handle subscriptions there. The events are still implemented in testdriver:
```
...
promise_test(async (test) => {
    ...
    const log_entry_promise = test.events["log.entryAdded"].once();
    ...
}, "Assert log event is logged", {events: {"log.entryAdded": test_driver.bidi.log.entry_added}});
```
The `test_driver` is required to allow `testharness` managing the subscriptions. The user is supposed to use events via testharness (` test.events...`). The specific API can be discussed.
2. Make testdriver aware of testharness:
```
promise_test(async (test) => {
    ...
    await test_driver.bidi.log.entry_added.subscribe(test);
    ...
}, "Assert log event is logged");
```
This will not completely address the concern of testharness relying on events.
3. Do not unsubscribe from the events.

I find the option **1** is the preferable. WDYT?

File: rfcs/testdriver_bidi.md [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+`WKTRTestharnessExecutor` etc) implementations should be done in the same way in order to support WebDriver BiDi.
+
+###### Messages from testdriver to wptrunner
+
+The communication channel between testdriver and wptrunner is set in the `do_testharness` method. Currently the messages from testdriver to wptrunner it is done via long poll by calling BaseProtocolPart.execute_async_script which is synchronous (async in the name indicates that the script is wrapped in a special way before sent to the browser), and blocks processing the event  loop in wptrunner, which in turn blocks the event processing while waiting for the next message form testdriver to wptrunner. To overcome this limitation, this RFC changes the wptrunner - testdriver communication channel to asynchronous non-blocking  [`BidiScriptProtocolPart.async_call_function`](https://github.com/web-platform-tests/wpt/pull/44649/files#diff-e5a8911dd97e0352b1b26d8ce6ef0a92b25378f7c9e79371a1eb1b1834bc9a8dR761) if available.  `WebDriverBidiProtocol`’s event loop should be used for message processing. This behavioral change is done only for protocols supporting BiDi.
```
[the remark this answers is not in the document]
Comment (sadym-chromium) [reply rfcs/testdriver_bidi.md:139]: Is it `s/with the harness/with the testdriver`, or I misread the comment?
> ... but instead have the harness emit custom `script.message` events to communicate information **with the testdriver**...

Do you mean in case of BiDi, the testharness should still make a long poll for getting commands from testdriver, but emit `message` events directly to `window`, avoiding `__wptrunner_process_next_event`? I'm not sure if I'm getting the point.

File: rfcs/testdriver_bidi.md [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+
+##### Alternatives
+
+* [Extend `CallbackHandler` to support async actions](#extend-callbackhandler-to-support-async-actions).
+* [Extend testdriver - wptrunner transport with `async_action`](#extend-testdriver---wptrunner-transport-with-async_action).
+
+##### Introduce `AsyncCallbackHandler`
+
+Add a `AsyncCallbackHandler`, which is an extension of `CallbackHandler` with support for async actions. It requires an event loop, in which it will schedule the tasks. Async action processing is done in the same way as CallbackHandler is done, but in the dedicated task. This would not allow raising unexpected exceptions in wptrunner. Testdriver will still be notified about those unexpected exceptions. [Example](https://github.com/web-platform-tests/wpt/pull/44649/files#diff-3de439d55203876d452599a1a14968409df69e5869b16918ceec378a6b3345a4R813) implementation.
```
[the remark this answers is not in the document]
Comment (sadym-chromium) [reply rfcs/testdriver_bidi.md:110]: Actually, the `ActionContext` is not needed in the `AsyncCallbackHandler`, as the action can be run in any context. Updated the example.

File: rfcs/testdriver_bidi.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jgraham) [reply rfcs/testdriver_bidi.md:33]: I didn't really mean each `test()` in a page, but between top-level test files.

The problem is that if the test file is allowed unfettered access to BiDi event subscriptions we can't isolate it from any subscriptions that the harness itself depends on, and doing something like closing the bidi session after each test would be very high overhead. The most obvious example would be if we wanted to use `script.message` to drive the harness, but also allow scripts themselves to emit custom messages. But it could also happen that, for example, we wanted the harness to be able to subscribe to `console.log` messages and re-emit them in the harness logs. Allowing the test to unsubscribe from those events would break that. So probably the harness has to manage subscriptions and know which events it subscribed to, and which should be forwarded back to the browser.

File: rfcs/testdriver_bidi.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jgraham) [reply rfcs/testdriver_bidi.md:139]: s/with the harness/with the browser/ yes.

In general the current way the testdriver event queue works, with `__process_next_event` is a workaround for limitations in WebDriver classic. I've been assuming that once we have BiDi we'd be able to move away from that model of polling an event queue in a single window, and instead use the browser event loop directly, and communicate via WebDriver BiDi events rather than blocking script evaluation. 

We don't necessarily have to do that right away, but I'd at least like to see what the migration path looks like,

[the remark this answers is not in the document]
Comment (sadym-chromium) [reply rfcs/testdriver_bidi.md:139]: > Do you mean in case of BiDi, the testharness should still make a long poll for getting commands from testdriver, but emit `message` events directly to `window`, avoiding `__wptrunner_process_next_event`? I'm not sure if I'm getting the point.

Oh, I got it. You mean the harness can provide a `channel` for the client-side wptrunner. I guess it is possible and makes sense, but I'd prefer to consider it in a different RFC to reduce execution risks.

File: rfcs/testdriver_bidi.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (OrKoN) [reply rfcs/testdriver_bidi.md:33]: Would it be sufficient if we implement automatic clean up of events similar to how BiDi's own test suite does it https://github.com/web-platform-tests/wpt/blob/d7b9fab5017587079985d5ab29781fcf3043533e/webdriver/tests/support/fixtures_bidi.py#L50 ? we could have executewebdriver.py record all subscribe calls coming from the test file and run corresponding unsubscribe calls filtering out events required by the testharness? In that case, unsubscribe calls would not be exposed to testdriver. There will probably be some edge cases around unsubscribing context-specific and global subscriptions at the same time.

Alternative ideas would require changing the spec:

1) support multiple light-wight BiDi sessions that do not require starting/shutting down the browser.
2) returning a subscriptionId in subscribe calls so that only specific subscriptions could be cleaned up.

[the remark this answers is not in the document]
Comment (sadym-chromium) [reply rfcs/testdriver_bidi.md:33]: I believe the API in **option 1** partly addresses the concerns you mentioned. It allows implementing event subscription and filtering in the [testharness](https://github.com/web-platform-tests/wpt/blob/91e2e520993bdc9820a0ddd619d27242f3abba88/resources/testharness.js#L1027) without exposing that logic to the test file, but the specific filtering of let's say `script.message` intended to be send to the test or to the testharness can be tricky.

> So probably the harness has to manage subscriptions and know which events it subscribed to, and which should be forwarded back to the browser.

Do you mean the backend testharness, or the client-side?

[the remark this answers is not in the document]
Comment (jgraham) [reply rfcs/testdriver_bidi.md:33]: FWIW @OrKoN's option 2 is something we'd like to propose for the BiDi spec (although it might have to be opt-in to _require_ a handle to unsubscribe since that's not how things currently work).

File: rfcs/testdriver_bidi.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jgraham) [reply rfcs/testdriver_bidi.md:139]: If you have some time to prototype that I'd appreciate it. As mentioned one concern in general is that making the wrong first step could make it harder in the long run if we bake in assumptions that don't work for other use cases.

File: rfcs/testdriver_bidi.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (sadym-chromium) [reply rfcs/testdriver_bidi.md]: Updated the `AsyncCallbackHandler` section

File: rfcs/testdriver_bidi.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (sadym-chromium) [reply rfcs/testdriver_bidi.md:33]: The proper long-term solution would be to provide separate sessions for the testharness and for each test files. This would provide proper isolations and tear down.

As an intermediate solution, in the new revision of the [example PR](https://github.com/web-platform-tests/wpt/pull/44649/files#diff-e5a8911dd97e0352b1b26d8ce6ef0a92b25378f7c9e79371a1eb1b1834bc9a8dR128) I implemented clean up on the `wptrunner` side as a part of `BidiEventsProtocolPart`. This allowed to remove `unsubscribe` from the testdriver API. We can always add it, if we have a valid use case.

@jgraham WDYT?

File: rfcs/testdriver_bidi.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (OrKoN) [reply rfcs/testdriver_bidi.md]: Since the communication between the testharness and the browser is internal and it is not exposed to test authors, it looks like it can be addressed separately. What assumptions from this RFC could prevent us from changing the event implementation in the future? 

It sounds like a preload script + a channel for setting up biderectional communication would not require API changes on the testdriver.js side.

File: rfcs/testdriver_bidi.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (sadym-chromium) [reply rfcs/testdriver_bidi.md]: I updated the RFC and the example with the new version. The protocol part `WebDriverBidiEventsProtocolPart` has a method `unsubscribe_all`, which is responsible for keeping track of all the subscriptions.

File: README.md [hunk 5]
```diff
@@ -21,7 +21,7 @@ or users of web-platform-tests. Examples of where it is likely to be useful:
 Cases where the RFC process need *not* be used:
 
  - Introducing a new top-level directory for a new specification.
- - Extending testdriver.js with a method that closely matches a [WebDriver](https://w3c.github.io/webdriver/) endpoint. To notify maintainers of testdriver.js vendor integration, label the pull request `testdriver.js`. (This exemption was introduced by [RFC 127](https://github.com/web-platform-tests/rfcs/blob/master/rfcs/rfc_exemption_testdriver_method.md).)
+ - Extending testdriver.js with a method that closely matches a [WebDriver Classic](https://w3c.github.io/webdriver/) or [WebDriver BiDi](https://w3c.github.io/webdriver-bidi) endpoints. To notify maintainers of testdriver.js vendor integration, label the pull request `testdriver.js`. (This exemption was introduced by [RFC 127](https://github.com/web-platform-tests/rfcs/blob/master/rfcs/rfc_exemption_testdriver_method.md) and [RFC 185](https://github.com/web-platform-tests/rfcs/blob/master/rfcs/testdriver_bidi.md).)
```
Comment (OrKoN) [inline README.md:24]:
```suggestion
 - Extending testdriver.js with a method that closely matches a [WebDriver Classic](https://w3c.github.io/webdriver/) endpoint or a [WebDriver BiDi](https://w3c.github.io/webdriver-bidi) command or event. To notify maintainers of testdriver.js vendor integration, label the pull request `testdriver.js`. (This exemption was introduced by [RFC 127](https://github.com/web-platform-tests/rfcs/blob/master/rfcs/rfc_exemption_testdriver_method.md) and [RFC 185](https://github.com/web-platform-tests/rfcs/blob/master/rfcs/testdriver_bidi.md).)
```

**Comment by sadym-chromium** [comment]
> I'd also like to see examples of how we could use the API for use cases that currently aren't possible or are prohibitively difficult, for example running actions in cross-origin windows. BiDi should unlock that, since it adds the ability to message arbitrary contexts. However the focus on just getting console.log events doesn't really reveal if this design is a useful stepping stone to that outcome. I worry that choices here could impede our ability to make the best use of BiDi in the future, if tests end up relying on accidental features of the design.

The core issue with handling several browsing context from a testdriver is in the fact the testdriver is not aware of browsing context ids. I see two approaches to address this:
1. When BiDi is available, allow testdriver to use window itself instead of the `window.__wptrunner_id` as a `context`. On the testdriver side, parse the testdriver messages and extract the browsing context ID.
2. Make testdriver be aware of it's own browsing context ID. This can be done by one of this ways:
    1. Add to `script.addPreloadScript` a new argument type `browsingContext`, which will be resolved to the actual browsingContext ID when called.
    2. Set a promise `window.__browsing_context_id` via `script.addPreloadScript`, and resolve it via `script.evaluate`.

**Comment by sadym-chromium** [comment]
> I'd also like to see examples of how we could use the API for use cases that currently aren't possible or are prohibitively difficult, for example running actions in cross-origin windows. BiDi should unlock that, since it adds the ability to message arbitrary contexts. However the focus on just getting console.log events doesn't really reveal if this design is a useful stepping stone to that outcome. I worry that choices here could impede our ability to make the best use of BiDi in the future, if tests end up relying on accidental features of the design.

I updated the [example](https://github.com/web-platform-tests/wpt/pull/44649/files#diff-e5a8911dd97e0352b1b26d8ce6ef0a92b25378f7c9e79371a1eb1b1834bc9a8dR757) by adding a parameter `contexts` to [`subscribe`](https://github.com/web-platform-tests/wpt/pull/44649/files#diff-9d6d3e870d033b8bc4d75b388fcc11dcf332e0293c286b2d57734c74e22020f7R42) action, which can be either a window proxy or a string representing browsing context id. This became possible because of the `_get_next_message` uses BiDi when it is available.

**Comment by sadym-chromium** [comment]
@jgraham do you think we need to extend other test in the example to make a more explicit use case of BiDi?

File: rfcs/testdriver_bidi.md [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+```javascript
+promise_test(async () => {
+    const some_message = "SOME MESSAGE";
+    // Subscribe to `log.entryAdded` BiDi events in the current window. This will not add a listener to the page.
+    await test_driver.bidi.log.entry_added.subscribe({contexts: [window]});
+    // Add a listener for the log.entryAdded event. This will not subscribe to the event, so the subscription is
+    // required before. The cleanup is done automatically after the test is finished.
+    let on_log_entry;
+    const log_entry_promise = new Promise(r => on_log_entry = r);
+    test_driver.bidi.log.entry_added.on(on_log_entry);
```
Comment (OrKoN) [inline rfcs/testdriver_bidi.md:237]:
```suggestion
    const log_entry_promise = new Promise(resolve => test_driver.bidi.log.entry_added.on(resolve));
```

File: rfcs/testdriver_bidi.md [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+```javascript
+promise_test(async () => {
+    const some_message = "SOME MESSAGE";
+    // Subscribe to `log.entryAdded` BiDi events. This will not add a listener to the page.
+    await test_driver.bidi.log.entry_added.subscribe();
+    // Add a listener for the log.entryAdded event. This will not subscribe to the event, so the subscription is
+    // required before. The cleanup is done automatically after the test is finished.
+    let on_log_entry;
+    const log_entry_promise = new Promise(r => on_log_entry = r);
+    test_driver.bidi.log.entry_added.on(on_log_entry);
```
Comment (OrKoN) [inline rfcs/testdriver_bidi.md:37]:
```suggestion
    const log_entry_promise = new Promise(resolve => test_driver.bidi.log.entry_added.on(resolve));
```

File: rfcs/testdriver_bidi.md [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+}, "Assert log event is logged");
+```
+
+### Required changes
+
+The proposal changes the following parts:
+
+#### Testdriver
+
+The testdriver requires support of bidi actions, aevent processing, and exposing it as an API.
```
Comment (OrKoN) [inline rfcs/testdriver_bidi.md:55]:
```suggestion
The testdriver requires support of bidi actions, event processing, and exposing it as an API.
```

**Comment by sadym-chromium** [comment]
> I'd also like to see examples of how we could use the API for use cases that currently aren't possible or are prohibitively difficult, for example running actions in cross-origin windows. BiDi should unlock that, since it adds the ability to message arbitrary contexts. However the focus on just getting console.log events doesn't really reveal if this design is a useful stepping stone to that outcome. I worry that choices here could impede our ability to make the best use of BiDi in the future, if tests end up relying on accidental features of the design.

Here is a draft example of what it can look like: https://github.com/sadym-chromium/wpt/pull/450

File: rfcs/testdriver_bidi.md [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+`WKTRTestharnessExecutor` etc) implementations should be done in the same way in order to support WebDriver BiDi.
+
+###### Messages from testdriver to wptrunner
+
+The communication channel between testdriver and wptrunner is set in the `do_testharness` method. Currently the messages from testdriver to wptrunner is done via long poll by calling `BaseProtocolPart.execute_async_script` which is synchronous (`async` in the name indicates that the script is wrapped in a special way before sent to the browser), and blocks processing the event loop in wptrunner, which in turn blocks the event processing while waiting for the next message form testdriver to wptrunner. To overcome this limitation, this RFC changes the wptrunner - testdriver communication channel to asynchronous non-blocking  [`BidiScriptProtocolPart.async_call_function`](https://github.com/web-platform-tests/wpt/pull/44649/files#diff-e5a8911dd97e0352b1b26d8ce6ef0a92b25378f7c9e79371a1eb1b1834bc9a8dR761) if available.  `WebDriverBidiProtocol`’s event loop should be used for message processing. This behavioral change is done only for protocols supporting BiDi.
```
Comment (jgraham) [inline rfcs/testdriver_bidi.md:135]: `async_call_function` seems like it could just be `call_function` since everything in bidi is async? That would also match the spec better.

Comment (sadym-chromium) [reply rfcs/testdriver_bidi.md]: Renames

**Comment by foolip** [comment]
I'll go ahead and merge this now. The RFC has been open for over a month and it's fair to say that we've ensured "that all participants have time to consider it." If anyone from Apple wants to leave feedback on the design, it's not too late for that even after implementation has begun.
